src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/CompilerScope.cs

```diff
@@ -169,10 +169,10 @@ internal CompilerScope Exit()
 
         internal void EmitVariableAccess(LambdaCompiler lc, ReadOnlyCollection<ParameterExpression> vars)
         {
-            if (NearestHoistedLocals != null)
+            if (NearestHoistedLocals != null && vars.Count > 0)
             {
                 // Find what array each variable is on & its index
-                var indexes = new List<long>(vars.Count);
```

Comment:
This turned out to be an unnecessary allocation when there are no variables. We created the list anyway and then went on to check the length of the list which gets populated in a loop with fixed iteration count and no conditionals inside, only to conclude it was throw-away work. Restructuring the conditionals here.